.mage/end_to_end.go

```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		return err
+	}
+	if (changed && !isCI) || !dumpExists {
+		os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+		mg.SerialDeps(Dev.DBErase, Dev.InitStack, EndToEnd.DBDump)
+	} else {
+		mg.SerialDeps(EndToEnd.DBRestore)
+	}
+	return nil
+}
+
+func (endToEnd EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.BuildIfNecessary)
+}
+
+// Prepare prepares the server for running end to end tests.
+func (endToEnd EndToEnd) Prepare() {
+	mg.Deps(endToEnd.prepareDB, endToEnd.prepareBuild)
+}
+
+// DBDump performs a database dump to the .cache folder.
+func (EndToEnd) DBDump() error {
+	if mg.Verbose() {
+		fmt.Printf("Execute database dump\n")
+	}
+	_ = os.Mkdir(".cache", 0755)
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", testDatabaseName, "--insecure")
+}
+
+// DBRestore restores the database using a previously generated dump.
+func (EndToEnd) DBRestore() error {
+	mg.Deps(Js.Deps, Dev.DBStart)
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
+	}
+	n, err := node()
+	if err != nil {
+		return err
+	}
+	return n("./.mage/restore-db-dump.js", "--db", testDatabaseName)
+}
+
+// Cypress runs the Cypress end-to-end tests.
+func (endToEnd EndToEnd) Cypress() error {
+	mg.Deps(endToEnd.WaitUntilReady)
+	if mg.Verbose() {
+		fmt.Println("Running Cypress E2E Tests")
+	}
+	cypress, err := endToEnd.cypress()
+	if err != nil {
+		return err
+	}
+	return cypress("run")
+}
+
+// CypressOpen runs the Cypress end-to-end tests in interactive mode.
+func (endToEnd EndToEnd) CypressOpen() error {
+	if mg.Verbose() {
+		fmt.Println("Running Cypress in interactive mode")
+	}
+	cypress, err := endToEnd.cypress()
+	if err != nil {
+		return err
+	}
+
+	return cypress("open")
+}
+
+// StartTestStack starts TTS in end-to-end test configuration.
+func (endToEnd EndToEnd) StartTestStack() error {
+	mg.Deps(Js.BuildIfNecessary)
+	if mg.Verbose() {
+		fmt.Println("Starting stack in end-to-end test configuration")
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	_, err := outputGo("run", "./cmd/ttn-lw-stack", "start")
+	return err
+}
+
+// WaitUntilReady waits until the web endpoints become available. For CI use.
+func (endToEnd EndToEnd) WaitUntilReady() error {
+	if mg.Verbose() {
+		fmt.Println("Waiting for the stack to be ready...")
+	}
+	for i := 0; i < 100; i++ {
+		resp, _ := http.Get(getTestURL() + "/oauth")
+		if resp != nil && resp.StatusCode == 200 {
+			return nil
+		}
+		time.Sleep(1000 * time.Millisecond)
+	}
+	return errors.New("Could not connect to server")
+}
+
+// RunFrontend starts the frontend end-to-end tests from scratch.
+func (endToEnd EndToEnd) RunFrontend() {
+	if mg.Verbose() {
+		fmt.Println("Running end-to-end frontend based tests")
+	}
+	mg.Deps(EndToEnd.Cypress)
+}
+
+// Run starts the end-to-end tests from scratch.
+func (endToEnd EndToEnd) Run() {
```

Comment:
What's the point of this mage target? Why not just use `RunFrontend`?

Reply:
The end-to-end test must not all be frontend-based, later on, we might have CLI based end-to-end tests as well. This target setup accounts for that.

Reply:
The keyword is *might*.
We will add it when we need it.